CleverRaven/Cataclysm-DDA
# Inventory view update - Show color for food in containers, show worst rot status for containers with multiple items

#### Summary

SUMMARY: Interface "Add Blue color for inventory display of food inside sealed containers + Show 'glass jar > [Color of rot status]pickles'"



#### Purpose of change

These changes are for inventory interaction convenience, with four changes made as outlined below:
1) [EDIT: This one will be added later when items have a reference to their parent container!] Due to request, added the ability to show food that is in a sealed container (and therefore not rotting) in a separate color in both inventory and [E] Eat dialogue boxes. Light blue was selected for being somewhat similar to cyan (for other comestibles) while being a bit more readable than Blue.
[i] Inventory:
![image](https://user-images.githubusercontent.com/55260779/103143376-96f5e400-46c9-11eb-8b59-f640cd81fb27.png)
[E] Eat dialogue:
![image](https://user-images.githubusercontent.com/55260779/103143381-af65fe80-46c9-11eb-8563-bf2c45c1e25e.png)
2) [EDIT: The number will also have to be added separately so that ammo won't have a buggy / double display] Added the number of stacks of an item inside of a container (if it's the only item in a container), ie.:
```glass jar > pickles (1)```
3) Added a different coloring for the second half of a container with an item if it's food to reflect its perishable status, such as:
![image](https://user-images.githubusercontent.com/55260779/103143032-6d39be80-46c3-11eb-8a2d-20ad3bbaab1e.png)
4) Same was added for a container with multiple items, taking the worst rot state (see cargo pants, where 25 items is brown since one of the items inside is rotten; messenger bag did not contain food at 1 level in) - NOTE this only goes 1-layer deep!:
![image](https://user-images.githubusercontent.com/55260779/103143083-a9215380-46c4-11eb-96e9-ff3f336fb19a.png)



#### Describe the solution

1) Updates to the nc_color color_in_inventory() function were made to handle locating an item's container and sealed status. The visit_items() function was used to search surrounding squares and the player's inventory for a given item's container.
2) An update to the tname() function allowed outputting the number of stacks of a contained item.
3) An update to the same section of the tname() function as in (2) was made to allow insertion of <color_COLOR></color> tags into the second half of an item description for container contents.
4) An update to the next section of the tname() function after (3) was made to show the worst rot state of multiple items within a container. Priority is as follows:
```Comestible (Nonperishable) < Perishable < Old < Rotten```
All of the above are superceded by Sealed
Note that allergy / uncooked / nausea status are not shown when there are more than 1 item in the container, although support for those can be added as well.



#### Describe alternatives you've considered

- Directly putting the <color_COLOR> indicator in the [E] Eat dialogue code of game_inventory around the "sealed" descriptor would be possible, although this will not color the name of the item itself, and also doesn't color the item in any other inventory screens.
- Tried blue instead of light blue as the sealed color, it's somewhat more difficult to read:
![image](https://user-images.githubusercontent.com/55260779/103143145-fa7e1280-46c5-11eb-87dc-06f2087202c4.png)
- Tried to avoid having to search every surrounding square and player's inventory, but directly calling visit_items in the color_in_inventory() function isn't legal, and the item itself was found not to contain the parent container anyway when forcing visit_items to work by creating an item_location referencing the item, or by force-removing the const typing of the referenced item.



#### Testing

Testing the coloring - Load the game:
1) Spawn 2x Canned Chicken ([f] contained)
2) Drop one of the cans on the floor
3) Open [i] to check coloring in inventory
4) Open [/] and check [a] all neighboring squares to check coloring of neighboring items
5) Open [E] to check coloring within Eat menu

Sample of how the [E] Eat dialogue currently looks:
![image](https://user-images.githubusercontent.com/55260779/103143192-4b8e0680-46c6-11eb-9e48-2f2a24f80e5f.png)

Same of the [/] Advanced Inventory dialogue:
![image](https://user-images.githubusercontent.com/55260779/103143205-5e084000-46c6-11eb-90e8-b4799565e1bc.png)
**NOTE** Notice that when a container with an item is highlighted in adv inventory, the second half doesn't get highlighted, although this doesn't happen in [i] inventory view (?), not sure if there's an easy fix for this. They also do not properly grey out when in the inactive window.

Testing multiple items in containers:
1) Spawn a fillet of fish 
2) Spawn wheat
3) Spawn butter
4) Pass time until the fillet of fish's status changes to old or rotten
5) Spawn a small cardboard box
6) Add all items to cardboard box and take them out progressively to ensure the coloring of the "> # items" text is correct.
Cardboard box with a rotten item:
![image](https://user-images.githubusercontent.com/55260779/103143296-14205980-46c8-11eb-8fae-152b378cac0d.png)
Cardboard box with at most a perishable item:
![image](https://user-images.githubusercontent.com/55260779/103143299-2f8b6480-46c8-11eb-9c3b-fd13258639e1.png)



#### Additional context

The display bug with the advanced inventory dialogue (see testing section's "NOTE" in bold) might need fixing, suggestions are welcome.
